model/common/src/icon4py/model/common/grid/icon.py

```diff
@@ -164,20 +168,23 @@ def n_shift(self):
     def lvert_nest(self):
         return True if self.config.lvertnest else False
 
-    def get_start_index(self, dim: gt_common.Dimension, marker: int) -> gt_builtins.int32:
+    def start_index(self, domain: h_grid.Domain):
         """
         Use to specify lower end of domains of a field for field_operators.
 
         For a given dimension, returns the start index of the
         horizontal region in a field given by the marker.
         """
-        return self.start_indices[dim][marker]
+        if domain.local:
+            # special treatment because this value is not set properly in the underlying data.
+            return 0
```

Comment:
Does it make sense we add an assert here to make sure the index does not go below zero and above number of cells?

Reply:
wouldn't it be better then at lines 94-95 where the arrays `_start_indeces` `_end_indices` are filled?

Reply:
Two things can be out of bounds: the index to access `start_index[CellDim]` which is an array of size 14, that the indices we use to access it (these are the indices set to the `Domain`s and it is checked for being between [0, 14) in `horizontal.py::._map_to_index `

Then again the values in `start_index` and `end_index` can be invalid Field indices (for example -1 as we had it... ) or something larger than the number of cells. That I agree generall with Jacopo should be checked when setting them. However it is not so easy to fill them. It needs knowledge of the domain decomposition and the refin_ctrl values. And we have and will have several sources of these data (DWD and MPI gridfile, that do it differently, serialized data...) So I chose to make all doubtful entries in these arrays inaccessible by using these predefined `Zone` and for this one case where we access it and we know it might be false to ignore the index in the array all together.

Reply:
Also I am currently working on exactly reconstructing these arrays for decomposed grid, so there I might fix it, but there might be other sources of these data in the future and in the case of the single node run, I would simply read what is in the gridfile (which is less wrong than what ICON does) the -1 the add it in the code.

Reply:
Long story short, we could add an assert here, but would still have to bypass the `start_index` array in this case, in order to be able to run at all and not always just get an `AssertionError